### PR TITLE
feat/ota-3196/Report aktualizr unit test results on gitlab

### PR DESCRIPTION
**Pull request by Zee314159**
Draft code. Just to trigger a CI test.

### CODE REVIEW

**Comment by Zee314159** [comment]
This failure `/aktualizr/CONTRIBUTING.md:41: error: unable to resolve reference to 'CHANGELOG.md' for \ref command (warning treated as error, aborting now)`  seems not caused by me. The need xml files have been generated, still waiting for GitLab administrator to enable the `junit_pipeline_view` feature, but I'm not sure if this is the right solution.

File: src/libaktualizr/uptane/uptane_network_test.cc [hunk 1]
```diff
@@ -203,7 +203,7 @@ TEST(UptaneNetwork, LogConnectivityRestored) {
   http->connectSwitch = true;
   testing::internal::CaptureStdout();
   result = up->fetchMeta();
-  EXPECT_EQ(result.status, result::UpdateStatus::kUpdatesAvailable);
+  EXPECT_EQ(result.status, result::UpdateStatus::kError);
```
Comment (patrickvacek) [inline src/libaktualizr/uptane/uptane_network_test.cc]: This doesn't seem right. Was something going wrong here?

**Comment by patrickvacek** [comment]
> This failure `/aktualizr/CONTRIBUTING.md:41: error: unable to resolve reference to 'CHANGELOG.md' for \ref command (warning treated as error, aborting now)` seems not caused by me.

Agreed. Perhaps a rebase will help, since I think this was a doxygen issue that Laurent recently fixed.

> The needed xml files have been generated, still waiting for GitLab administrator to enable the `junit_pipeline_view` feature, but I'm not sure if this is the right solution.

Looks like it could work! Do you have an example or a link to the generated output so we can take a look?

File: src/libaktualizr/uptane/uptane_network_test.cc (hunk 1 above)
Comment (Zee314159) [reply src/libaktualizr/uptane/uptane_network_test.cc]: Nothing's wrong. Just make a failure case to test the GitLab report feature.

**Comment by Zee314159** [comment]
> Do you have an example or a link to the generated output so we can take a look?

https://main.gitlab.in.here.com/olp/edge/ota/connect/client/aktualizr/-/jobs/1080837/artifacts/browse/build-coverage/Testing/20191219-0609/
`junit_pipeline_view` feature has been enabled, but the there's no report showed. I added `junit` in docker files, it's all the same, not sure what's going wrong.

**Comment by Zee314159** [comment]
https://na-stage.gitlab.in.here.com/olrudenk/temp_traffic_mobile_application/pipelines/69/test_report 
This project works fine on gitlab-stg.

File: docker/Dockerfile.ubuntu.xenial [hunk 2]
```diff
@@ -22,6 +22,7 @@ RUN apt-get update && apt-get -y install --no-install-suggests --no-install-reco
   gcc \
   git \
   jq \
+  junit \
```
Comment (patrickvacek) [inline docker/Dockerfile.ubuntu.xenial]: I think you should only need to add `junit` for the bionic image. I don't think it'll be used by the xenial or Debian testing images.

File: cmake-modules/CodeCoverage.cmake [hunk 3]
```diff
@@ -156,7 +156,7 @@ function(SETUP_TARGET_FOR_COVERAGE_LCOV)
         COMMAND ${LCOV_PATH} --gcov-tool ${GCOV_PATH} -c -i -d . -o ${Coverage_NAME}.base
 
         # Run tests
-        COMMAND ${Coverage_EXECUTABLE}
+        COMMAND ${Coverage_EXECUTABLE} ${Coverage_EXECUTABLE_ARGS}
```
Comment (lbonn) [inline cmake-modules/CodeCoverage.cmake]: Curious why is this required?

**Comment by lbonn** [comment]
I think the problem is that the xml output from ctest is not compatible with junit and needs to be converted. See for example this cmake issue https://gitlab.kitware.com/cmake/cmake/issues/18654.

File: docker/Dockerfile.ubuntu.xenial (hunk 2 above)
Comment (lbonn) [reply docker/Dockerfile.ubuntu.xenial]: Actually, do we need junit at all on the images?

Comment (Zee314159) [reply docker/Dockerfile.ubuntu.xenial]: Yes, it's a long shot... seems not needed, tried it anyway.

File: cmake-modules/CodeCoverage.cmake (hunk 3 above)
Comment (Zee314159) [reply cmake-modules/CodeCoverage.cmake]: `set(CTEST_EXTRA_ARGS -T Test --no-compress-output)`
`setup_target_for_coverage_lcov(NAME coverage EXECUTABLE ${CMAKE_CTEST_COMMAND} ${CTEST_EXTRA_ARGS} USES_TERMINAL)`
from `aktualizr/tests/CMakeLists.txt` the `Test.xml` is only generated when this argument is set.

**Comment by Zee314159** [comment]
> I wish it didn't involve pulling in php, though.

Yes, It looks php didn't do anything to help. I'll remove it see if it still works.

**Comment by Zee314159** [comment]
> I wish it didn't involve pulling in php, though.
> 
> Yes, It looks php didn't do anything to help. I'll remove it see if it still works.

**Comment by Zee314159** [comment]
> I wish it didn't involve pulling in php, though.
> 
> Yes, It looks php didn't do anything to help. I'll remove it see if it still works.

**Comment by Zee314159** [comment]
> I wish it didn't involve pulling in php, though.
> 
> Yes, It looks php didn't do anything to help. I'll remove it see if it still works.

Fixed.

**Comment by Zee314159** [comment]
> Cool, glad to see this simplified. Where did the .xsl file come from? We need to be sure the copyright is friendly.

https://gitlab.kitware.com/cmake/cmake/issues/18654 
We need to add a description in `thirdparty.spdx` right?

**Comment by patrickvacek** [comment]
> https://github.com/manticoresoftware/manticoresearch/tree/master/misc/junit
> We need to add a description in `thirdparty.spdx` right?

Yep. It's GPL, but since it's only used for CI, I think it's fine. You'll also need to add an entry to `.ort.yml` to exclude it.
